src/service/position-observer-impl.js

```diff
@@ -0,0 +1,413 @@
+        this.entries_.splice(i, 1);
+        if (this.entries_.length == 0) {
+          this.stopCallback();
+        }
+        return;
+      }
+    }
+    dev().error(TAG, 'cannot unobserve unobserved element');
+  }
+
+  /** @param {!Object} unusedEntry */
+  updateEntryPosition_(unusedEntry) {}
+
+  startCallback() {}
+
+  stopCallback() {}
+
+  increaseFidelityCallback() {}
+
+  decreaseFidelityCallback() {}
+}
+
+export class AmpDocPositionObserver extends AbstractPositionObserver {
+
+  /** @param {!./ampdoc-impl.AmpDoc} ampdoc */
+  constructor(ampdoc) {
+
+    super(ampdoc);
+
+    /** @private {boolean} */
+    this.inScroll_ = false;
+
+    /** @private {boolean} */
+    this.measure_ = false;
+
+    /** @private {number} */
+    this.scrollTimer_ = Date.now();
+
+    /** @private {Array<function()>} */
+    this.unlisteners_ = [];
+  }
+
+  /* @override */
+  startCallback() {
+    // listen to viewport scroll event to help pass determine if need to
+    this.unlisteners_.push(this.viewport_.onScroll(() => {
+      this.inScroll_ = true;
+      this.scrollTimer_ = Date.now();
+      this.schedulePass_();
+      setTimeout(() => {
+        if (Date.now() - this.scrollTimer_ < 500) {
+          // viewport scroll in the last 500 ms, wait
+          return;
+        }
+        // assume scroll stops, if no scroll event in the last 500,
+        this.inScroll_ = false;
+      }, 500);
+    }));
+    this.unlisteners_.push(this.viewport_.onChanged(() => {
+      this.vsync_.measure(() => {
+        this.pass_(true);
+      });
+    }));
+    // QQQ: Already call this.updateEntryPosition_(entry) in observe()? Needed?
+    // initial pass
+    this.vsync_.measure(() => {
+      this.pass_(true);
+    });
+  }
+
+  /* @override */
+  stopCallback() {
+    while (this.unlisteners_.length) {
+      const unlisten = this.unlisteners_.pop();
+      unlisten();
+    }
+  }
+
+  /** @param {boolean=} recursive */
+  schedulePass_(recursive) {
+    // TODO (@zhouyx, #9208):
+    // P1: account for effective fidelity using this.effectiveFidelity
+    // P2: do passes on onDomMutation (if available using MutationObserver)
+    if (!recursive && this.measure_) {
+      // call of schedulePass_ from viewport onScroll
+      // Do nothing if currently measure with calling schedulePass recursively
+      return;
+    }
+    this.measure_ = true;
+    if (!this.inScroll_) {
+      // not in scroll, do not need to measure
+      this.measure_ = false;
+      return;
+    }
+    this.vsync_.measure(() => {
+      this.pass_();
+      this.schedulePass_(true);
+    });
+  }
+
+  /**
+   * @param {boolean=} force
+   * @private
+  */
+  pass_(force) {
+    for (let i = 0; i < this.entries_.length; i++) {
+      const entry = this.entries_[i];
+      if (!force && entry.turn != 0) {
+        // Not ready for their turn yet.
+        entry.turn--;
+        continue;
+      }
+      // Reset entry.turn value.
+      if (!force) {
```

Comment:
Sorry, I thought the starting value was just `LOW_FEDELITY_FRAME_COUNT`. So now we have two points:

1. Why are we setting to a random number? These won't all be scheduled on the same frame, so randomness just adds complexity without any benefit.
2. My original comment was why we're not resetting if we just measured.